conbench/entities/run.py

```diff
@@ -532,7 +593,22 @@ class _RunFacadeSchemaCreate(marshmallow.Schema):
     error_type = marshmallow.fields.String(
         required=False, metadata={"description": field_descriptions["error_type"]}
     )
-    github = marshmallow.fields.Nested(GitHubCreate(), required=False)
+    github = marshmallow.fields.Nested(
+        SchemaGitHubCreate(),
+        required=False,
+        metadata={
+            "description": conbench.util.dedent_rejoin(
+                """
+                Use this object to tell Conbench with which commit the
+                Run/BenchmarkResult is associated.
+
+                This field can be left out for testing purposes. Note however
```

Comment:
Another important use case here is when you run benchmarks on a state that doesn't exist on the GitHub-known git graph, like when something runs in CI on a transient merge-commit between the HEAD of a PR and HEAD of the default branch, or when you run benchmarks locally after making changes that you haven't pushed. In those cases, you must leave this field out, or risk polluting your data.

Reply:
> on a state that doesn't exist on the GitHub-known git graph,

Thanks for pointing this out.

Asking naively: what's the purpose of submitting such a result to Conbench? To _later_ add commit information?

Reply:
An example: we're currently setting up "pre-merge" benchmarks on one of our company's internal projects. These run in CI on pull requests, building the project and running benchmarks on the "commit" that the CI checks out.

In most CI systems, when you run CI on a PR commit, it actually runs on a temporary merge-commit between the HEAD of the PR branch and the HEAD of main, and that temporary commit doesn't actually exist on the GitHub-known git graph. For example, this happens when using GitHub Actions; check out [the docs](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request). The GITHUB_REF is `refs/pull/:prNumber/merge`, which is briefly explained by [this SO answer](https://stackoverflow.com/a/63595981).

This sort of goes against our [documented strategy](https://conbench.github.io/conbench/pages/benchmarking_and_the_dev_process.html#possible-alternatives-and-their-limitations) of always benchmarking on the explicit PR head, even if the PR branch is forked from an older default-branch commit. We're benchmarking on a state of code that may not exist or be reference-able in the future, and that's bad for reproducibility. So we shouldn't connect it to the canonical git graph.

(@jonkeane has told me about a brief time years ago where we _did_ post the commit hash of the "transient" merge-commit to Conbench, and it was a nightmare to maintain and explain to users. It's simply hard to draw conclusions on transient history.)

So our strategy will be to post these pre-merge runs with no git commit information, since they don't belong on the git graph. We won't be able to view plots for those results, but we will at least have a URL to point to for looking at their raw data and errors. And importantly, we should still be able to hit `/api/compare/runs/some_good_run_id...pre_merge_run_id` to see whether there were regressions compared to some "good" baseline run. That latter use case is why I've been working on issues like #384 lately.

Reply:
> So our strategy will be to post these pre-merge runs with no git commit information, since they don't belong on the git graph.

ack

> To later add commit information?

just summarizing, the answer is "no"

> we will at least have a URL to point to for looking at their raw data and errors [...] able to hit /api/compare/runs/some_good_run_id...pre_merge_run_id to see whether there were regressions compared to some "good" baseline run. That latter use case is why I've been working on issues like https://github.com/conbench/conbench/issues/384 lately.

OK, two specific 'value' items. Thanks!

Reply:
> Should change this text again.

https://github.com/conbench/conbench/pull/1425